pkg/controller/util/federated_informer.go

```diff
@@ -176,79 +173,78 @@ func NewFederatedInformer(
 		return data
 	}
 
-	federatedInformer.clusterInformer.store, federatedInformer.clusterInformer.controller = cache.NewInformer(
-		&cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
-				return fedClient.CoreV1alpha1().FederatedClusters(namespaces.FederationNamespace).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return fedClient.CoreV1alpha1().FederatedClusters(namespaces.FederationNamespace).Watch(options)
-			},
-		},
+	DeleteFunc := func(old interface{}) {
+		oldCluster, ok := old.(*fedv1a1.FederatedCluster)
+		if ok {
+			var data []interface{}
+			if clusterLifecycle.ClusterUnavailable != nil {
+				data = getClusterData(oldCluster.Name)
+			}
+			federatedInformer.deleteCluster(oldCluster)
+			if clusterLifecycle.ClusterUnavailable != nil {
+				clusterLifecycle.ClusterUnavailable(oldCluster, data)
+			}
+		}
+	}
+
+	AddFunc := func(cur interface{}) {
+		curCluster, ok := cur.(*fedv1a1.FederatedCluster)
+		if !ok {
+			glog.Errorf("Cluster %v/%v not added; incorrect type", curCluster.Namespace, curCluster.Name)
+		} else if IsClusterReady(curCluster) {
+			federatedInformer.addCluster(curCluster)
+			glog.Infof("Cluster %v/%v is ready", curCluster.Namespace, curCluster.Name)
+			if clusterLifecycle.ClusterAvailable != nil {
+				clusterLifecycle.ClusterAvailable(curCluster)
+			}
+		} else {
+			glog.Infof("Cluster %v/%v not added; it is not ready.", curCluster.Namespace, curCluster.Name)
+		}
+	}
+
+	UpdateFunc := func(old, cur interface{}) {
+		oldCluster, ok := old.(*fedv1a1.FederatedCluster)
+		if !ok {
+			glog.Errorf("Internal error: Cluster %v not updated.  Old cluster not of correct type.", old)
+			return
+		}
+		curCluster, ok := cur.(*fedv1a1.FederatedCluster)
+		if !ok {
+			glog.Errorf("Internal error: Cluster %v not updated.  New cluster not of correct type.", cur)
+			return
+		}
+		if IsClusterReady(oldCluster) != IsClusterReady(curCluster) || !reflect.DeepEqual(oldCluster.Spec, curCluster.Spec) || !reflect.DeepEqual(oldCluster.ObjectMeta.Annotations, curCluster.ObjectMeta.Annotations) {
+			var data []interface{}
+			if clusterLifecycle.ClusterUnavailable != nil {
+				data = getClusterData(oldCluster.Name)
+			}
+			federatedInformer.deleteCluster(oldCluster)
+			if clusterLifecycle.ClusterUnavailable != nil {
+				clusterLifecycle.ClusterUnavailable(oldCluster, data)
+			}
+
+			if IsClusterReady(curCluster) {
+				federatedInformer.addCluster(curCluster)
+				if clusterLifecycle.ClusterAvailable != nil {
+					clusterLifecycle.ClusterAvailable(curCluster)
+				}
+			}
+		} else {
+			glog.V(4).Infof("Cluster %v not updated to %v as ready status and specs are identical", oldCluster, curCluster)
+		}
+	}
+
+	var err error
+	federatedInformer.clusterInformer.store, federatedInformer.clusterInformer.controller, err = NewGenericInformerWithEventHandler(
```

Comment:
Why is it desirable to change the code to accept the event handler functions separately?

Reply:
reverted. ptal